src/nvim/eval.c

```diff
@@ -9442,22 +9442,71 @@ static void f_getcmdwintype(typval_T *argvars, typval_T *rettv)
   rettv->vval.v_string[0] = cmdwin_type;
 }
 
-/*
- * "getcwd()" function
- */
+/// `getcwd('s')` function
+///
+/// @pre The type of the argument `s` must be a string or nothing.
 static void f_getcwd(typval_T *argvars, typval_T *rettv)
 {
-  char_u      *cwd;
+  assert(argvars->v_type == VAR_STRING || argvars->v_type == VAR_UNKNOWN);
+
+  char_u *cwd = NULL;
+  char_u *scp = argvars->vval.v_string;  // scope
 
   rettv->v_type = VAR_STRING;
   rettv->vval.v_string = NULL;
+
   cwd = xmalloc(MAXPATHL);
-  if (os_dirname(cwd, MAXPATHL) != FAIL) {
-    rettv->vval.v_string = vim_strsave(cwd);
+  cwd[0] = '\0';  // initialize to empty string
+
+  // If there is no argument return the current working directory.
+  // This is the same behaviour as in Vim.
+  //
+  // Also if global is requested and the current working directory is the
+  // global one.
+  if (argvars->v_type == VAR_UNKNOWN || (!globaldir && scp[0] == 'g')) {
+    if (os_dirname(cwd, MAXPATHL) != FAIL) {
+      goto done;
+    } else {
+      goto end;
+    }
+  }
+
+  char_u *from = NULL;  // The original string to copy
+
+  switch (scp[0]) {  // only check the first character
+    case 'g':  // global
+      from = globaldir;
+      break;
+    case 't':  // tab
+      from = curtab->localdir;
+      break;
+    case 'w':  // window
+      from = curwin->w_localdir;
+      break;
+    default:
+      EMSG("E740: valid arguments are either 't' or 'w' or nothing.");
+      break;
+  }
+
+  if (!from) {
+    goto done;
+  }
+  // Copy the `from` string
+  for (int i = 0; i < MAXPATHL; ++i) {
+    cwd[i] = from[i];
+    if (from[i] == '\0') {
+      goto done;
+    }
+  }
+  // If we made it through the loop the string was too large for copying
+  goto end;
+
+done:
+  rettv->vval.v_string = vim_strsave(cwd);
```

Comment:
Here we could use `cwd` instead of `vim_strsave(cwd)`. No need to duplicate the string again.
The `goto end;` further up could be replaced with an error message and `xfree(cwd);return;`.